tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala

```diff
@@ -150,4 +168,72 @@ object OutputManager {
       val rundir = Files.createTempDirectory(outDir, s"${specName}_${nicedate}T${nicetime}_")
       runDirOpt = Some(rundir)
     }
+
+  /** Create a PrintWriter to the file formed by appending `fileParts` to the `base` file */
+  def printWriter(base: File, fileParts: String*): PrintWriter = {
+    val file = fileParts.foldLeft(base)((file, part) => new File(file, part))
+    new PrintWriter(new FileWriter(file))
+  }
+
+  /** Create a PrintWriter to the file formed by appending `fileParts` to the `base` file */
+  def printWriter(base: Path, fileParts: String*): PrintWriter = {
+    printWriter(base.toFile(), fileParts: _*)
+  }
+
+  /**
+   *  Create a PrintWriter to the file formed by appending `fileParts` to the `base` file
+   *
+   * E.g., to create a writer to the file `foo/bar/bas.json`:
+   *
+   *    val w = printWriter("foo", "bar", "baz.json")
+   */
+  def printWriter(base: String, fileParts: String*): PrintWriter = {
+    printWriter(Paths.get(base), fileParts: _*)
+  }
+
+  /** Apply f to the writer w, being sure to close w */
+  private def withWriter(f: PrintWriter => Unit)(w: PrintWriter) = {
+    try {
+      f(w)
+    } finally {
+      w.close()
+    }
+  }
+
+  def withWriterToFile(file: File)(f: PrintWriter => Unit): Unit = {
+    withWriter(f)(printWriter(file))
+  }
+
+  /** Applies `f` to a PrintWriter created by appending the `parts` to the `runDir` */
+  def withWriterRelativeToRunDir(parts: String*)(f: PrintWriter => Unit): Unit = {
+    val w = printWriter(runDirOpt.get, parts: _*)
+    withWriter(f)(w)
+  }
+
+  /**
+   * Conditionally applies a function to a PrintWriter constructed relative to the intermediate directory
+   *
+   * @param parts path parts describing a path relative to the intermediate directory (all parents must exist)
+   * @param f a function that will be applied to the `PrintWriter`, if the `IntermediateFlag` is set.
+   * @return `true` if the `IntermediateFlag` is true, and `f` can be applied to the PrintWriter
+   *        created by appending the `parts` to the intermediate output dir. Otherwise, `false`.
+   */
+  def withWriterRelativeToIntermediateDir(parts: String*)(f: PrintWriter => Unit): Boolean = {
+    if (!flags(Names.IntermediateFlag)) {
+      false
+    } else {
+      intermediateDirFile match {
+        case Some(dir) => {
+          withWriter(f)(printWriter(dir, parts: _*))
+          true
+        }
+        case None => {
+          val dirName = intermediateDirPathOpt.getOrElse("")
+          logger.error(
+              s"Unable to find or create intermediate output directory ${dirName}. Intermediate output will not be written.")
+          false
+        }
```

Comment:
@konnov I've opted for logging an error rather than crashing in the event of failure. The rationale being that if you're in the middle of a long running checking session, you may not want to lose all your progress and results just because this file cannot be created for some reason.

Does that seem sensible, or should it rather raise an exception?

It just occurs to me, it would be better if we just ensure all the directories needed for the configured output exist at startup, then error out early if they don't. However, I think that reworking of the initialization of the `OutputManager` is probably out of scope for this PR, and should instead be included in #1069. How does that sound?

Reply:
I think it's a good compromise.